### DISCUSSION TITLE
Question: Does buildah depend on qemu-user-static?

### DISCUSSION THREAD

**Discussion by lazywhite** [containers/buildah#4736] [General]
It seems like when installing buildah, there is no need to install qemu-user-static any more? 
1. Is qemu-user-static required in older version? if it is, since what version  this dependency get removed?
2. if not depend on it, how buildah can build multiarch images?
3. if we run buildah inside a docker container, are there extra requirements to build multiarch images?

**Comment by flouthoc** [comment]
I think `qemu-user-static` is still needed inorder to build `multi-arch` images but @nalind could confirm this much better.

**Comment by nalind** [comment]
I'm not aware of us ever having listed `qemu-user-static` as a dependency in binary packages, but it's what I'd recommend if your Dockerfile uses RUN instructions and you're building for a platform that your CPU can't natively run code for, i.e., when using the `--platform` option to build for something other than the local architecture.
If RUN is being used, last I checked, you needed `qemu-user-static` or something like it installed.  The packaging would configure the kernel to use the appropriate qemu binary as an interpreter when it's asked to run binaries that aren't marked in their binary headers as having been built for a natively-supported architecture.
If your Dockerfile doesn't use RUN instructions, though, the build won't cause buildah to try to run any such binaries, so the kernel won't have an opportunity or reason to return an "exec format error" error for them.
Regardless of whether or not the non-native-architecture binary is being run inside of a container, `qemu-user-static` needs to be installed on a host, since it's the host's kernel that needs to be configured to call out to the right interpreter when anything tries to run a binary that isn't built for a natively-supported architecture.

**Reply by lazywhite** [reply]
if buildah does  depend on qemu-user-static, I think it should be in installation doc, and give a explaination

**Reply by lazywhite** [reply]
maybe you guys also need to write a doc about how to run buildah inside docker container

**Reply by rhatdan** [reply]
Buildah does not depend on qemu-user-static.  Buildah runs executables, if the kernel does not support the arch of the executable, then the executable will fail, or you can install an emulation tool to tell the kernel how to translate the code to be able to run on the native arch.

**Reply by lazywhite** [reply]
alright, the conclusion is "if users want to build multiarch images and contains RUN command, users should install "qemu-user-static" base on their OS"

**Reply by rhatdan** [reply]
Yes, Buildah does not have any magic way to make this happen.  :^)
